runtime/components/jacMachine/include/features/nodeModules.hpp

```diff
@@ -148,6 +154,284 @@ class NodeModuleLoader {
+        duk_push_string(ctx, id);
+        duk_put_prop_string(ctx, -2, "\xff" "moduleId");
+
+        /* require.cache */
+        duk_push_global_stash(ctx);
+        (void) duk_get_prop_string(ctx, -1, "\xff" "requireCache");
+        duk_put_prop_string(ctx, -3, "cache");
+        duk_pop(ctx);
+
+        /* require.main */
+        duk_push_global_stash(ctx);
+        (void) duk_get_prop_string(ctx, -1, "\xff" "mainModule");
+        duk_put_prop_string(ctx, -3, "main");
+        duk_pop(ctx);
+    }
+
+    static duk_ret_t dukHandleRequire(duk_context *ctx) {
+        /*
+        *  Value stack handling here is a bit sloppy but should be correct.
+        *  Call handling will clean up any extra garbage for us.
+        */
+
+        const char *id;
+        const char *parent_id;
+        duk_idx_t module_idx;
+        duk_idx_t stash_idx;
+        duk_int_t ret;
+
+        duk_push_global_stash(ctx);
+        stash_idx = duk_normalize_index(ctx, -1);
+
+        duk_push_current_function(ctx);
+        (void) duk_get_prop_string(ctx, -1, "\xff" "moduleId");
+        parent_id = duk_require_string(ctx, -1);
+        (void) parent_id;  /* not used directly; suppress warning */
+
+        /* [ id stash require parent_id ] */
+
+        id = duk_require_string(ctx, 0);
+
+        (void) duk_get_prop_string(ctx, stash_idx, "\xff" "modResolve");
+        duk_dup(ctx, 0);   /* module ID */
+        duk_dup(ctx, -3);  /* parent ID */
+        duk_call(ctx, 2);
+
+        /* [ ... stash ... resolved_id ] */
+
+        id = duk_require_string(ctx, -1);
+
+        if (dukGetCachedModule(ctx, id)) {
+            goto have_module;  /* use the cached module */
+        }
+
+        dukPushModuleObject(ctx, id, false);
+        dukPutCachedModule(ctx);  /* module remains on stack */
+
+        /*
+        *  From here on out, we have to be careful not to throw.  If it can't be
+        *  avoided, the error must be caught and the module removed from the
+        *  require cache before rethrowing.  This allows the application to
+        *  reattempt loading the module.
+        */
+
+        module_idx = duk_normalize_index(ctx, -1);
+
+        /* [ ... stash ... resolved_id module ] */
+
+        (void) duk_get_prop_string(ctx, stash_idx, "\xff" "modLoad");
+        duk_dup(ctx, -3);  /* resolved ID */
+        (void) duk_get_prop_string(ctx, module_idx, "exports");
+        duk_dup(ctx, module_idx);
+        ret = duk_pcall(ctx, 3);
+        if (ret != DUK_EXEC_SUCCESS) {
+            dukDelCachedModule(ctx, id);
+            (void) duk_throw(ctx);  /* rethrow */
+        }
+
+        if (duk_is_string(ctx, -1)) {
+            duk_int_t ret;
+
+            /* [ ... module source ] */
+            ret = duk_safe_call(ctx, dukEvalModuleSource, NULL, 2, 1);
+            if (ret != DUK_EXEC_SUCCESS) {
+                dukDelCachedModule(ctx, id);
+                (void) duk_throw(ctx);  /* rethrow */
+            }
```

Comment:
Odsud je stále varianta `dukEvalModuleSource` která kopírouje source a má ho na stacku, ale používá se to jen při `require` v JS kódu, a tam zatím neposkytujeme to načítání z FS, vypadá, že to si dělá duktape sám(?).